pkg/cache/sql/db/transaction/transaction.go

```diff
@@ -0,0 +1,94 @@
+/*
+Package transaction provides a client for a live transaction, and interfaces for some relevant sql types. The transaction client automatically performs rollbacks  on failures.
+The use of this package simplifies testing for callers by making the underlying transaction mock-able.
+*/
+package transaction
+
+import (
+	"context"
+	"database/sql"
+
+	"github.com/pkg/errors"
+)
+
+// Client provides a way to interact with the underlying sql transaction.
+type Client struct {
+	sqlTx  SQLTx
+	closed bool
+}
+
+// SQLTx represents a sql transaction
+type SQLTx interface {
+	Exec(query string, args ...any) (sql.Result, error)
+	Stmt(stmt *sql.Stmt) *sql.Stmt
+	Commit() error
+	Rollback() error
+}
+
+// Stmt represents a sql stmt. It is used as a return type to offer some testability over returning sql's Stmt type
+// because we are able to mock its outputs and do not need an actual connection.
+type Stmt interface {
+	Exec(args ...any) (sql.Result, error)
+	Query(args ...any) (*sql.Rows, error)
+	QueryContext(ctx context.Context, args ...any) (*sql.Rows, error)
+}
+
+// NewClient returns a Client with the given transaction assigned.
+func NewClient(tx SQLTx) *Client {
+	return &Client{sqlTx: tx}
+}
+
+// Commit commits the transaction and then unlocks the database.
+func (c *Client) Commit() error {
+	c.closed = true
+	return c.sqlTx.Commit()
+}
+
+// Exec uses the sqlTX Exec() with the given stmt and args. The transaction will be automatically rolled back if Exec()
+// returns an error.
+func (c *Client) Exec(stmt string, args ...any) error {
+	_, err := c.sqlTx.Exec(stmt, args...)
```

Comment:
Agreed we had a problem here.

The only purpose of closed was to avoid a double rollback on Close, but that had a race condition possibility as access to close was not serialized.

Fixed in 12a491e